### PR TITLE
feat: Enable internal connections for JMX port.

### PR DESCRIPTION
**Pull request by julianGoh17**
Created a new field in the Kafka CRD which would allow you to enable the
JMX metrics on the kafka broker. When the field is specified, the kafka
broker starts up with port 9404 for the JMX. Then it exposes 9404 on the
kafka headless service to be able to collect the metrics.

Signed-off-by: Julian Goh <juliangohsl@gmail.com>

### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

_Please describe your pull request_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaJmxRemote.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
+@JsonPropertyOrder({"secure"})
+@EqualsAndHashCode
+public class KafkaJmxRemote implements UnknownPropertyPreserving, Serializable {
+    private static final long serialVersionUID = 1L;
+    private boolean isSecure;
+
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
+
+    @Description("Description for whether jmx remote options are enabled")
```
Comment (tomaley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaJmxRemote.java]: Toggle security for the JMX port

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java [hunk 2]
```diff
@@ -371,8 +373,14 @@ public static Kafka createKafkaCluster(String clusterCmNamespace, String cluster
...
         kafkaClusterSpec.setReadinessProbe(livenessProbe);
         if (metricsCm != null) {
-            kafkaClusterSpec.setMetrics(metricsCm);
+            kafkaClusterSpec.setPrometheusMetrics(metricsCm);
+        }
+        KafkaJmxRemote jmxOptions = new KafkaJmxRemote();
+        if (isJmxEnabled) {
+
         }
+        kafkaClusterSpec.setKafkaJmxRemote(jmxOptions);
```
Comment (tomaley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java]: Should be inside the above `if` block

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java [hunk 3]
```diff
@@ -140,6 +141,7 @@
 
     private static boolean openShift;
     private static boolean metrics;
+    private static boolean jmxSecured;
```
Comment (tomaley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: jmxSecure

File: docker-images/kafka/scripts/kafka_run.sh [hunk 4]
```diff
@@ -25,6 +25,33 @@ fi
...
+  else
+    # expose the port insecurely
+    KAFKA_JMX_OPTS="${KAFKA_JMX_OPTS} -Dcom.sun.management.jmxremote.authenticate=false"
+
+fi
+
+KAFKA_OPTS="${KAFKA_OPTS} ${KAFKA_JMX_OPTS}"
+
 # enabling Prometheus JMX exporter as Java agent
 if [ "$KAFKA_METRICS_ENABLED" = "true" ]; then
```
Comment (tomaley) [inline docker-images/kafka/scripts/kafka_run.sh]: Rename to Prometheus metrics

File: documentation/assemblies/assembly-kafka-jmx-remote.adoc [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+// Save the context of the assembly that is including this one.
+// This is necessary for including assemblies in assemblies.
+// See also the complementary step on the last line of this file.
+
+[id='assembly-jmx-options-{context}']
+
+= JMX Remote
+
+{ProductName} supports obtaining JMX Metrics on the Kafka brokers. JMX metrics can be used to obtain various metrics
+specific to each Kafka brokers such as `BytesPerSecond` that a broker is using. {ProductName} supports opening a password
```
Comment (tomaley) [inline documentation/assemblies/assembly-kafka-jmx-remote.adoc]: broker

File: documentation/assemblies/assembly-kafka-jmx-remote.adoc [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+and username protected JMX port or an non-protected JMX port. The JMX port on the kafka brokers can be enabled with security which will create a username and password protected JMX
```
Comment (tomaley) [inline documentation/assemblies/assembly-kafka-jmx-remote.adoc]: a non-....

Comment (tomaley) [inline documentation/assemblies/assembly-kafka-jmx-remote.adoc]: Capitalise Kafka, remove last sentence, an -> a, brokers -> broker, remove that a broker is using

File: documentation/modules/ref-kafka-jmx-remote-options.adoc [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+In the above example, the JMX port on the Kafka brokers are opened and then the JMX port on the Kafka headless service
+is opened. This will allow you to deploy a pod internally into a cluster and obtain JMX metrics by using the headless
+service and specifying which broker you want to address. For example, to get metrics from broker 0 we address the
+headless service by doing the following:  "_`<cluster-name>`_-kafka-0- _`<cluster-name>`_-_`<headless-service-name>`_" on the JMX port.
+
+By enabling a secure option, a secret will be generated containing the username and password that the Kafka
+brokers will start up with. That means that all Kafka brokers will use the same username and password to protect the
+JMX port. The username and password that is used to secure the port are stored in the "_`<cluster-name>`_-kafka-jmx" secret.
+
+You can get JMX metrics from the Kafka brokers by deploying a pod that can obtain JMX metrics into the cluster containing
```
Comment (tomaley) [inline documentation/modules/ref-kafka-jmx-remote-options.adoc]: Already said this in the above text?

File: documentation/modules/ref-kafka-jmx-remote-options.adoc [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  kafka:
+    # ...
+    jmxRemote:
+      secure: false
+    # ...
+  zookeeper:
+    # ...
+----
+
+This will just create the headless service and you can follow a similar approach as described above to deploy a pod into
```
Comment (tomaley) [inline documentation/modules/ref-kafka-jmx-remote-options.adoc]: Will not create the headless service, will add another port to the headless service...

File: documentation/modules/ref-kafka-jmx-remote-options.adoc (hunk 7 above)
Comment (julianGoh17) [reply documentation/modules/ref-kafka-jmx-remote-options.adoc]: I agree

File: install/cluster-operator/050-Deployment-strimzi-cluster-operator.yaml [hunk 9]
```diff
@@ -71,6 +71,8 @@ spec:
           value: strimzi/kafka-bridge:latest
         - name: STRIMZI_LOG_LEVEL
           value: INFO
+        - name: STRIMZI_IMAGE_PULL_POLICY
+          value: Never
```
Comment (tomaley) [inline install/cluster-operator/050-Deployment-strimzi-cluster-operator.yaml]: Wrong

**Pull request merged**